ocaml/xapi/xapi_host_driver.ml

```diff
@@ -0,0 +1,140 @@
+   it under the terms of the GNU Lesser General Public License as published
+   by the Free Software Foundation; version 2.1 only. with the special
+   exception on linking described in file LICENSE.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Lesser General Public License for more details.
+ *)
+
+module D = Debug.Make (struct let name = __MODULE__ end)
+
+open D
+module Unixext = Xapi_stdext_unix.Unixext
+
+let ( // ) = Filename.concat
+
+let invalid_value field value =
+  raise Api_errors.(Server_error (invalid_value, [field; value]))
+
+let internal_error fmt =
+  Printf.ksprintf
+    (fun msg ->
+      error "%s" msg ;
+      raise Api_errors.(Server_error (internal_error, [msg]))
+    )
+    fmt
+
+let create ~__context ~host ~name_label ~versions ~active_version
+    ~selected_version =
+  info "%s: %s" __FUNCTION__ name_label ;
+  let ref = Ref.make () in
+  let uuid = Uuidx.to_string (Uuidx.make ()) in
+  Db.Host_driver.create ~__context ~ref ~uuid ~host ~name_label ~versions
+    ~active_version ~selected_version
+
+let destroy ~__context ~self =
+  D.debug "Destroying driver %s" (Ref.string_of self) ;
+  Db.Host_driver.destroy ~__context ~self
+
+(*
+* `/lib/modules/<kernel-version>/updates/` is searched by the kernel for
+  drivers - this is `selected_dir`.
+* The hierarchy is expected to contain symbolic links to the file
+  actually containing the driver in
+  `/lib/modules/<kernel-version>/xenserver/<driver>/<version>/<name>.ko` -
+  the base directory here is `all_versions_dir`. *)
+type driver_paths = {selected_dir: string; all_versions_dir: string}
+
+let get_drivers_base_path ~__context ~host =
+  let linux_version =
+    List.assoc "linux" (Db.Host.get_software_version ~__context ~self:host)
+  in
+  let base_dir = "/lib/modules" // linux_version in
+  {
+    selected_dir= base_dir // "updates"
+  ; all_versions_dir= base_dir // "xenserver"
+  }
+
+let reload_driver_selection ~__context ~host =
+  let linux_version =
+    List.assoc "linux" (Db.Host.get_software_version ~__context ~self:host)
+  in
+  try
+    let _, _ =
+      Forkhelpers.execute_command_get_output "/usr/sbin/depmod"
+        [
+          "-ae"
+        ; "-F"
+        ; Format.sprintf "/boot/System.map-%s" linux_version
+        ; linux_version
+        ]
+    in
+    let _, _ =
+      Forkhelpers.execute_command_get_output "/usr/sbin/dracut"
+        [
+          "-f"; Format.sprintf "/boot/initrd-%s.img" linux_version; linux_version
+        ]
+    in
+    ()
+  with Forkhelpers.Spawn_internal_error (err, _, _) ->
+    internal_error "Failed to reload driver selection (%s): %s" __FUNCTION__ err
+
+(* Runs on the necessary host *)
+let select ~__context ~self ~version =
+  let versions = Db.Host_driver.get_versions ~__context ~self in
```

Comment:
This is part of the first commit, which says "empty implementation", but it definitely isn't empty ;)